### PR TITLE
Crash Detection

### PR DESCRIPTION
**Pull request by samzhang5469**
## MPU Sensor
The MPU6050 is used to measure and calculate the state of the bike. Then the wireless module publishes the information to the MQTT broker under the /v3/wireless_module/3/mpu_data topic for acceleration, rotation and gyroscopic data, and /v3/wireless_module/3/crashed_state for crashed state.

## Setup
1. Follow the Getting Started instructions [here](https://github.com/monash-human-power/data-acquisition-system/tree/master/wireless_modules) 
2. Once you have created a local version of the config.py file you can navigate to the ```back_module``` folder.
3. Execute the upload script with 
```bash
# change this to the serial port of the ESP32
./upload.sh --port /dev/ttyUSB0
```
``` bash
# if this is taking a while in your shell, switch shells or alternatively manually upload these files
(
    "main.py" 
    "../config.py" 
    "../wireless_module.py"
    "../boot.py"
    "../mqtt_client.py"
    "../battery_reader.py"
    "../sensors/sensor_base.py"
    "../sensors/co2_sensor.py"
    "../sensors/dht_sensor.py"
    "../sensors/gps_sensor.py"
    "../sensors/mpu_sensor.py"
    "../sensors/reed_sensor.py"
    "../status_led.py"
    "../libraries/abc.py"
    "../libraries/MQ135/mq135.py"
    "../libraries/micropyGPS/micropyGPS.py"
    "../libraries/MPU6050-ESP8266-MicroPython/mpu6050.py"
) 
# using 
ampy -p /dev/ttyUSB0 -b 115200 put <FILE>
```
4. Verify that the file system is uploaded correctly using
```bash
# change /dev/ttyUSB0 to the serial port of the ESP32
ampy -p /dev/ttyUSB0 ls
```
5. Install and setup [Mosquitto](https://mosquitto.org/download/) and make sure its running

## Steps to Test

1. Switch on the primary display
3. Create a mobile hotspot with the previously used ESSID and password
5. Verify that the wireless module, primary display and subscribed devices are all connected to the hotspot
7. Interact with the wireless module through your computer (see [this](https://github.com/monash-human-power/data-acquisition-system/tree/master/wireless_modules/middle_module#steps-to-test))

## Current Tuning Method
It is a good idea to calibrate the sensors on before using it. Notably aim to calibrate the sensor if: the back wireless module is accidently dropped, or: after real crash has been detected.
In the current implementation of the accelerometer, you will need to manually calibrate the sensor if the readings have drifted. Parameters are stored as class attributes of the ```MpuSensor``` within ```mpu_sensor.py```. These tunable values include:
| Parameter| Description |
| --- | ----------- |
| ```xyz_calib_offset``` | Is used to ensure axis readings are zero while perpendicular to the direction of gravity. To tune this parameter. </br>1. Modify the mpu_sensor.py file by adding ```print(accel_values) ``` in line 80 </br>2. Lay the accelerometer with z-axis facing up, and ensure that the accelerometer in on a level surface with a leveler and that there is as little vibration as possible </br>3. Test the device and read the values of x and z. </br> 4. Pick values ```a,b``` for the other two axis x,y such that ```x+a≈0``` and ```y+b≈0```  <br> 5. Verify these values by reorienting the accelerometer with other axis facing directly up. </br> 5. Update the class attribute and reflash the wireless module again before moving on to the next parameter |
| ```xyz_calib_factor``` | This determines how sensitive the accelerometer is. Once xyz_calib_offset it tuning sufficiently, </br>1. Place the accelerometer on a level surface with z axis facing up, find the value of ```a``` such that ```z*a≈1``` </br> 2. Repeat this by reorienting the accelerometer with the calibration axis facing up and solving for ```a``` once again for all other axis readings </br>3. Update the parameter and reflash the wireless module|
| ```gyro_calib_offset``` | This parameter helps to offset the initial error in the gyroscope. Once xyz_calib_factor is complete </br> 1. Place MPU on a flat surface with as little vibration as possible, modify the ```mpu_sensor.py``` code to include ```print(gyro_values)``` instead on ```line 80```  and reflash the wireless module</br> 2. Pick values for ```gyro_calib_offset``` such that the printed values are as close to 0 as possible while at rest </br>3. Reflash the wireless module with the new values.</br> A method will be added to calibrate this automatically soon.|
| complemenatary ratio | The complementary ratio is not store as a class attribute. The tuning is left to the judgement of the user. See [notion](https://www.notion.so/monashhumanpower/Wireless-module-crash-detection-5295c2b1b84e44f980fb0b6abb87bd92) for approach. The actual values are the hardcoded values found in ```line 74``` and ```line 75``` in ```mpu_sensor.py``` |

Currently there is no way to tune gyroscope scaling without the use of specialized testing equipment.

### CODE REVIEW

File: wireless_modules/sensors/mpu6500.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
+# Copyright (c) 2018-2020 Mika Tuupola
```
Comment (jgOhYeah) [inline wireless_modules/sensors/mpu6500.py]: Should this file (mpu6500.py) be included as a submodule seeing as it is from another git repository? (not sure how it affects licensing if included here).
https://github.com/tuupola/micropython-mpu9250

File: wireless_modules/sensors/mpu_sensor_6500.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
+from machine import I2C
```
Comment (jgOhYeah) [inline wireless_modules/sensors/mpu_sensor_6500.py]: Is this file needed? It looks to me like it is a backup / work in progress that doesn't seem to be used, but I could be wrong. (mpu_sensor.py is the one that the back wireless module imports).

**Review by jgOhYeah — CHANGES_REQUESTED** [review]
I haven't tested it, but looks good overall.
A couple of things that may need to be addressed or discussed:
- Importing the mpu6500 library as a submodule to keep the version control / lineage of the library in tact. Does the previous mpu6050 library need to be removed from the submodules list in that case?
- A couple of files that appear to be works in progress and can be removed?
- Did the line endings do something weird to edit what seems like every file? Is this something that needs to be cleaned up (Windows vs Linux settings on the git client)?

File: DAS/das/V3_fake_module.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
Comment (jgOhYeah) [inline DAS/das/V3_fake_module.py]: Is this in here on purpose or is it a remnant from a failed merge at some point?

File: DAS/das/V3_fake_module.py [hunk 4]
```diff
@@ -236,3 +237,252 @@ def on_message(client, userdata, msg):
     client.loop_start()
     start_publishing(client, args)
     client.loop_stop()
+=======
```
Comment (jgOhYeah) [inline DAS/das/V3_fake_module.py]: Also potentially from a failed merge.

File: DAS/poetry.lock [hunk 5]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
Comment (jgOhYeah) [inline DAS/poetry.lock]: Part of a failed merge?

File: DAS/poetry.lock [hunk 6]
```diff
@@ -476,3 +477,689 @@ typing-extensions = []
 urllib3 = []
 xlsxwriter = []
 zipp = []
+=======
```
Comment (jgOhYeah) [inline DAS/poetry.lock]: Part of a failed merge?

File: DAS/poetry.lock [hunk 7]
```diff
@@ -476,3 +477,689 @@ typing-extensions = []
...
+]
+urllib3 = [
+    {file = "urllib3-1.26.9-py2.py3-none-any.whl", hash = "sha256:44ece4d53fb1706f667c9bd1c648f5469a2ec925fcf3a776667042d645472c14"},
+    {file = "urllib3-1.26.9.tar.gz", hash = "sha256:aabaf16477806a5e1dd19aa41f8c2b7950dd3c746362d7e3223dbe6de6ac448e"},
+]
+xlsxwriter = [
+    {file = "XlsxWriter-3.0.3-py3-none-any.whl", hash = "sha256:df0aefe5137478d206847eccf9f114715e42aaea077e6a48d0e8a2152e983010"},
+    {file = "XlsxWriter-3.0.3.tar.gz", hash = "sha256:e89f4a1d2fa2c9ea15cde77de95cd3fd8b0345d0efb3964623f395c8c4988b7f"},
+]
+>>>>>>> ab2fd7f99f13543d1c56b4afcd45ccf5b4b92e56
```
Comment (jgOhYeah) [inline DAS/poetry.lock]: Part of a failed merge?

File: DAS/pyproject.toml [hunk 8]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
Comment (jgOhYeah) [inline DAS/pyproject.toml]: Part of a failed merge?

File: DAS/pyproject.toml [hunk 9]
```diff
@@ -22,3 +23,29 @@ pytest = "^7.1.3"
 [build-system]
 requires = ["poetry-core>=1.0.0"]
 build-backend = "poetry.core.masonry.api"
+=======
```
Comment (jgOhYeah) [inline DAS/pyproject.toml]: Part of a failed merge?

File: DAS/pyproject.toml [hunk 10]
```diff
@@ -22,3 +23,29 @@ pytest = "^7.1.3"
...
+rich = "^12.2.0"
+XlsxWriter = "^3.0.3"
+
+[tool.poetry.dev-dependencies]
+pytest = "^6.1.2"
+
+[build-system]
+requires = ["poetry>=0.12"]
+build-backend = "poetry.masonry.api"
+>>>>>>> ab2fd7f99f13543d1c56b4afcd45ccf5b4b92e56
```
Comment (jgOhYeah) [inline DAS/pyproject.toml]: Part of a failed merge?

File: utils/poetry.lock [hunk 11]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
Comment (jgOhYeah) [inline utils/poetry.lock]: Part of a failed merge?

File: utils/poetry.lock [hunk 12]
```diff
@@ -268,3 +269,281 @@ wcwidth = [
     {file = "wcwidth-0.1.9-py2.py3-none-any.whl", hash = "sha256:cafe2186b3c009a04067022ce1dcd79cb38d8d65ee4f4791b8888d6599d1bbe1"},
     {file = "wcwidth-0.1.9.tar.gz", hash = "sha256:ee73862862a156bf77ff92b09034fc4825dd3af9cf81bc5b360668d425f3c5f1"},
 ]
+=======
```
Comment (jgOhYeah) [inline utils/poetry.lock]: Part of a failed merge?

File: utils/poetry.lock [hunk 13]
```diff
@@ -268,3 +269,281 @@ wcwidth = [
...
+]
+six = [
+    {file = "six-1.14.0-py2.py3-none-any.whl", hash = "sha256:8f3cd2e254d8f793e7f3d6d9df77b92252b52637291d0f0da013c76ea2724b6c"},
+    {file = "six-1.14.0.tar.gz", hash = "sha256:236bdbdce46e6e6a3d61a337c0f8b763ca1e8717c03b369e87a7ec7ce1319c0a"},
+]
+wcwidth = [
+    {file = "wcwidth-0.1.9-py2.py3-none-any.whl", hash = "sha256:cafe2186b3c009a04067022ce1dcd79cb38d8d65ee4f4791b8888d6599d1bbe1"},
+    {file = "wcwidth-0.1.9.tar.gz", hash = "sha256:ee73862862a156bf77ff92b09034fc4825dd3af9cf81bc5b360668d425f3c5f1"},
+]
+>>>>>>> ab2fd7f99f13543d1c56b4afcd45ccf5b4b92e56
```
Comment (jgOhYeah) [inline utils/poetry.lock]: Part of a failed merge?

File: utils/pyproject.toml [hunk 14]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
Comment (jgOhYeah) [inline utils/pyproject.toml]: Part of a failed merge?

File: utils/pyproject.toml [hunk 15]
```diff
@@ -16,3 +17,23 @@ pytest = "^5.2"
 [build-system]
 requires = ["poetry>=0.12"]
 build-backend = "poetry.masonry.api"
+=======
```
Comment (jgOhYeah) [inline utils/pyproject.toml]: Part of a failed merge?

File: utils/pyproject.toml [hunk 16]
```diff
@@ -16,3 +17,23 @@ pytest = "^5.2"
...
+argparse = "^1.4.0"
+numpy = "^1.21.0"
+
+[tool.poetry.dev-dependencies]
+pytest = "^5.2"
+
+[build-system]
+requires = ["poetry>=0.12"]
+build-backend = "poetry.masonry.api"
+>>>>>>> ab2fd7f99f13543d1c56b4afcd45ccf5b4b92e56
```
Comment (jgOhYeah) [inline utils/pyproject.toml]: Part of a failed merge?

File: wireless_modules/config.example.py [hunk 17]
```diff
@@ -1,11 +0,0 @@
-ESSID = "<WIFI NETWORK NAME HERE>"
```
Comment (jgOhYeah) [inline wireless_modules/config.example.py]: Is this file meant to be deleted? It looks like it is meant to be kept as a reference to me (might have been renamed to make the actual config file at one point, which git may interpret as removing this one).

File: wireless_modules/wireless_module.py [hunk 18]
```diff
@@ -152,7 +179,7 @@ async def start_data_loop(self, interval):
         stop message - after which the process is repeated.
         :param interval: Integer representing number of seconds to wait before sending data.
         """
-        secs_to_ms = 1000
+        secs_to_ms = 200
```
Comment (jgOhYeah) [inline wireless_modules/wireless_module.py]: Should `secs_to_ms` be renamed to something like `time_scalar`... as 200ms isn't really 1 second (being a bit pedantic :) )?

File: wireless_modules_py/wind_module/config.py [hunk 19]
```diff
@@ -0,0 +1,6 @@
+MQTT_BROKER = "localhost"
```
Comment (jgOhYeah) [inline wireless_modules_py/wind_module/config.py]: Does this need to be turned into the `config.example.py` format that then gets copied to a gitignored `config.py` to make it a bit harder to accidently commit passwords? This is more of a question for the wind sensor project, but putting it here while I remember.

**Review by jgOhYeah — CHANGES_REQUESTED** [review]
Getting closer - good work.
Unfortunately it looks like there was a merge that failed at one point and added various bits of text for manual merging, which will break things at some point. These seem to have been added in commit  8a423ab.
For the more minor things that aren't necessarily specific to this PR / affect the functionality:
- A few questions about which config and example config files should be commited or removed.
- Naming of variables that suggest 1 second is 5 times faster than I thought it was :).

**Review by AroshHeenkenda — APPROVED** [review]
None of these changes should break anything, tested it out myself and works fine. PR approved Sam Zhang.

**Review by jgOhYeah — APPROVED** [review]
Looks good to me

**Pull request merged**